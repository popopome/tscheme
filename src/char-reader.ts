export interface ICharPos {
  pos: number;
  line: number;
  col: number;
}
export interface ICharReader {
  position(): ICharPos;
  current(): string;
  next(): string;
  seek(pos: number): void;
  getString(begin: ICharPos, end: ICharPos): string;
  getLineString(line: number): string;
}
export function makeCharReader(code: string): ICharReader {
  let pos = -1;
  let lines: number[] = [0];
  let col = 0;

  return {
    getLineString(line: number): string {
      if (line < lines.length) {
        const begin = lines[line];
        let end = line + 1 < lines.length ? lines[line + 1] : -1;
        if (end === -1) {
          end = begin;
          while (end < code.length) {
            if (code[end] === '\n') break;
            ++end;
          }
        }
        return code.substring(begin, end);
      }
      return '';
    },
    position() {
      return { pos, line: lines.length - 1, col };
    },
    current() {
      return code[pos];
    },
    next() {
      ++pos;
      if (pos >= code.length) return '';
      updateLineCol(pos);
      return this.current();
    },
    seek(position: number) {
      pos = position;
      for (let i = 0; i < lines.length; ++i) {
        if (pos < lines[i]) {
          lines.splice(0, i);
          updateLineCol(pos);
        }
      }
    },
    getString(begin: ICharPos, end: ICharPos) {
      return code.substring(begin.pos, end.pos);
    }
  };

  function updateLineCol(pos: number) {
    if (pos === 0) {
      lines = [0];
      col = 0;
    } else if (code[pos - 1] === '\n') {
      lines.push(pos);
      col = 0;
    } else {
      col = pos - lines[lines.length - 1];
    }
  }
}
