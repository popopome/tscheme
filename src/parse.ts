import { makeTokenizer, TokenType, Tokenizer, Token } from './tokenizer';
import { ICharPos, ICharReader } from './char-reader';

export interface Sym extends Token {
  type: TokenType.SYMBOL;
}
export interface Num extends Token {
  type: TokenType.NUMBER;
}
export interface Str extends Token {
  type: TokenType.STRING;
}

export type Atom = Sym | Num | Str;
export interface List {
  type: TokenType.LIST;
  begin: ICharPos;
  end: ICharPos;
  value: (Atom | List)[];
}

export type Exp = Atom | List;
export interface IAst {
  program: Exp;
}

function raiseParseError(msg: string, begin: ICharPos, end: ICharPos) {
  const err = new Error(msg);
  (err as any).tag = 'PARSE-ERROR';
  (err as any).msg = msg;
  (err as any).begin = begin;
  (err as any).end = end;
  throw err;
}

function parse2(tokenizer: Tokenizer): Exp {
  let tok = tokenizer.current();

  if (tok.type === TokenType.OPEN_PAREN) {
    const list: List = {
      type: TokenType.LIST,
      begin: tok.begin,
      end: tok.end,
      value: []
    };
    tokenizer.next();
    while (
      tokenizer.current().type !== TokenType.CLOSE_PAREN &&
      tokenizer.current().type !== TokenType.EOF
    ) {
      list.value.push(parse2(tokenizer));
    }
    if (tokenizer.current().type === TokenType.CLOSE_PAREN) {
      list.end = tokenizer.current().end;
      tokenizer.next();
      return list;
    }
    raiseParseError(
      'Unable to find close paren',
      list.begin,
      tokenizer.current().end
    );
  }

  if (tok.type === TokenType.CLOSE_PAREN) {
    raiseParseError('Unexpected close paren', tok.end, tok.end);
  }

  if (
    tok.type === TokenType.STRING ||
    tok.type === TokenType.SYMBOL ||
    tok.type === TokenType.NUMBER
  ) {
    tokenizer.next();
    return tok as Atom;
  }

  return null as any;
}

export function parse(reader: ICharReader): IAst {
  try {
    const tokenizer = makeTokenizer(reader);
    tokenizer.next();
    return {
      program: parse2(tokenizer)
    };
  } catch (err) {
    console.log(JSON.stringify(err));
    throw err;
  }
}
