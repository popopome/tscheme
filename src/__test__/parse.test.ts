import { parse } from '../parse';
import { makeCharReader } from '../char-reader';

test('parse should parse a given s-expression and return the AST', () => {
  expect(parse(makeCharReader('(a b)'))).toMatchSnapshot('SIMPLE-PARSE');
  expect(parse(makeCharReader('(a "thanks" 123 b)'))).toMatchSnapshot('LIST');
});
