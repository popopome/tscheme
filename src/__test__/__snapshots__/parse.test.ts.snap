// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse should parse a given s-expression and return the AST: LIST 1`] = `
Object {
  "program": Object {
    "begin": Object {
      "col": 0,
      "line": 0,
      "pos": 0,
    },
    "end": Object {
      "col": 17,
      "line": 0,
      "pos": 18,
    },
    "type": "LIST",
    "value": Array [
      Object {
        "begin": Object {
          "col": 1,
          "line": 0,
          "pos": 1,
        },
        "end": Object {
          "col": 2,
          "line": 0,
          "pos": 2,
        },
        "type": "SYMBOL",
        "value": "a",
      },
      Object {
        "begin": Object {
          "col": 3,
          "line": 0,
          "pos": 3,
        },
        "end": Object {
          "col": 11,
          "line": 0,
          "pos": 11,
        },
        "type": "STRING",
        "value": "thanks",
      },
      Object {
        "begin": Object {
          "col": 12,
          "line": 0,
          "pos": 12,
        },
        "end": Object {
          "col": 15,
          "line": 0,
          "pos": 15,
        },
        "type": "NUMBER",
        "value": 123,
      },
      Object {
        "begin": Object {
          "col": 16,
          "line": 0,
          "pos": 16,
        },
        "end": Object {
          "col": 17,
          "line": 0,
          "pos": 17,
        },
        "type": "SYMBOL",
        "value": "b",
      },
    ],
  },
}
`;

exports[`parse should parse a given s-expression and return the AST: SIMPLE-PARSE 1`] = `
Object {
  "program": Object {
    "begin": Object {
      "col": 0,
      "line": 0,
      "pos": 0,
    },
    "end": Object {
      "col": 4,
      "line": 0,
      "pos": 5,
    },
    "type": "LIST",
    "value": Array [
      Object {
        "begin": Object {
          "col": 1,
          "line": 0,
          "pos": 1,
        },
        "end": Object {
          "col": 2,
          "line": 0,
          "pos": 2,
        },
        "type": "SYMBOL",
        "value": "a",
      },
      Object {
        "begin": Object {
          "col": 3,
          "line": 0,
          "pos": 3,
        },
        "end": Object {
          "col": 4,
          "line": 0,
          "pos": 4,
        },
        "type": "SYMBOL",
        "value": "b",
      },
    ],
  },
}
`;
